### PR TITLE
Downloader with improved interaction with Sentry

### PR DESCRIPTION
**Pull request by mriccobene**
This PR aims at:

1. avoid exception condition due to gRPC failure (in some cases)

2. transfer some responsibility from BlockProvider & HeaderDownloader to SentryClient

### CODE REVIEW

File: node/silkworm/downloader/sentry_client.cpp [hunk 1]
```diff
@@ -18,11 +18,72 @@
...
 
+#include "rpc/SetStatus.hpp"
+#include "rpc/ReceiveMessages.hpp"
+
 namespace silkworm {
 
 SentryClient::SentryClient(std::string sentry_addr)
     : base_t(grpc::CreateChannel(sentry_addr, grpc::InsecureChannelCredentials())) {}
 
 void SentryClient::exec_remotely(SentryRpc& rpc) { base_t::exec_remotely(rpc); }
```
Comment (yperbasis) [inline node/silkworm/downloader/sentry_client.cpp:29]: `exec_remotely` is not virtual in `rpc::Client`; I don't see a point in defining it in `SentryClient`.

File: node/silkworm/downloader/sentry_client.hpp [hunk 2]
```diff
@@ -37,21 +40,38 @@ using SentryRpc = rpc::Call<sentry::Sentry>;
...
     void exec_remotely(SentryRpc& rpc);  // send a rpc request to the remote sentry
 
-    void stop() { stopping_.store(true); }
-    bool is_stopping() const { return stopping_.load(); }
+    enum Scope { BlockRequests = 0x01, BlockAnnouncements = 0x02, Other = 0x04 };
+    // enum values enable bit masking, for example: cope = BlockRequests & BlockAnnouncements
+
+    void subscribe(Scope, subscriber_t callback);  // subscribe with sentry to receive messages
+
+    /*[[long_running]]*/ void execution_loop();  // do a long-running loop to wait for messages
```
Comment (yperbasis) [inline node/silkworm/downloader/sentry_client.hpp:61]: Please mark with `override`.

File: node/silkworm/downloader/sentry_client.cpp [hunk 3]
```diff
@@ -18,11 +18,72 @@
 
 #include <silkworm/common/log.hpp>
 
+#include "rpc/SetStatus.hpp"
+#include "rpc/ReceiveMessages.hpp"
+
 namespace silkworm {
 
 SentryClient::SentryClient(std::string sentry_addr)
```
Comment (yperbasis) [inline node/silkworm/downloader/sentry_client.cpp:26]: Better pass `sentry_addr` by `const &`

File: node/silkworm/downloader/sentry_client.cpp (hunk 1 above)
Comment (AndreaLanfranchi) [reply node/silkworm/downloader/sentry_client.cpp:29]: Yes. See also the same [comment in previous PR](https://github.com/torquem-ch/silkworm/pull/483#issuecomment-965450396)

File: node/silkworm/downloader/sentry_client.cpp [hunk 4]
```diff
@@ -18,11 +18,72 @@
...
+            return SentryClient::Scope::BlockAnnouncements;
+        case sentry::MessageId::GET_BLOCK_HEADERS_66:
+        case sentry::MessageId::GET_BLOCK_BODIES_66:
+            return SentryClient::Scope::BlockRequests;
+        default:
+            return SentryClient::Scope::Other;
+    }
+}
+
+void SentryClient::subscribe(Scope scope, subscriber_t callback) { subscribers_[scope].push_back(callback); }
```
Comment (yperbasis) [inline node/silkworm/downloader/sentry_client.cpp:46]: .push_back(**std::move**(callback))

**Review by yperbasis — APPROVED** [review]

File: node/silkworm/downloader/sentry_client.cpp (hunk 1 above)
Comment (mriccobene) [reply node/silkworm/downloader/sentry_client.cpp:29]: The point is this: exec_remotely() is the most important method of the SentryClient so it's nice to see it in the SentryClient class declaration, for completeness and for easy of reading. Also the comment on the method completes the picture. 

But, anyway, I solved replacing the method redefinition with this directive:

`using base_t::exec_remotely; // exec_remotely(SentryRpc& rpc) sends a rpc request to the remote sentry`

File: node/silkworm/downloader/block_provider.hpp [hunk 5]
```diff
@@ -52,8 +42,19 @@ class BlockProvider : public ActiveComponent {  // an active component that must
     /*[[long_running]]*/ void execution_loop() override;  // main loop, receive messages from sentry and process them
 
   private:
-    void send_status();
-    void process_message(std::shared_ptr<InboundMessage> message);
+    using MessageQueue = ConcurrentQueue<std::shared_ptr<InboundMessage>>;  // used internally to store new messages
+
+    void receive_message(const sentry::InboundMessage&);
+
+    ChainIdentity chain_identity_;
```
Comment (yperbasis) [inline node/silkworm/downloader/block_provider.hpp:49]: `chain_identity_` is not used anywhere

File: node/silkworm/downloader/block_provider.cpp [hunk 6]
```diff
@@ -16,9 +16,13 @@
...
 
+#include <chrono>
+#include <thread>
+
 #include <silkworm/common/log.hpp>
 
 #include "internals/header_retrieval.hpp"
+#include "messages/InboundMessage.hpp"
 #include "rpc/ReceiveMessages.hpp"
 #include "rpc/SetStatus.hpp"
```
Comment (yperbasis) [inline node/silkworm/downloader/block_provider.cpp:27]: Unused include

File: node/silkworm/downloader/block_provider.hpp [hunk 7]
```diff
@@ -18,31 +18,21 @@
...
  * BlockProvider depends upon a SentryClient to connect to the remote sentry to receive requests and send responses.
  */
 class BlockProvider : public ActiveComponent {  // an active component that must run always
-
-    ChainIdentity chain_identity_;
-    Db::ReadOnlyAccess db_access_;
-    SentryClient& sentry_;
-
   public:
     BlockProvider(SentryClient& sentry, Db::ReadOnlyAccess db_access, ChainIdentity chain_identity);
```
Comment (yperbasis) [inline node/silkworm/downloader/block_provider.hpp:37]: Better pass `db_access` by `const &`

**Pull request merged**